pkg/config/remote/service/core_agent_stream.go

```diff
@@ -0,0 +1,145 @@
+package service
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+
+	"github.com/DataDog/datadog-agent/pkg/api/security"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/config/remote/service/tuf"
+	"github.com/DataDog/datadog-agent/pkg/proto/pbgo"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type coreAgentStream struct {
+	stream                       pbgo.AgentSecure_GetConfigUpdatesClient
+	streamCtx                    context.Context
+	cancel                       context.CancelFunc
+	streamCancel                 context.CancelFunc
+	agentClient                  pbgo.AgentSecureClient
+	mx                           sync.Mutex
+	currentConfigSnapshotVersion uint64
+	tufClient                    *tuf.DirectorPartialClient
+}
+
+func newCoreAgentStream() (*coreAgentStream, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	token, err := security.FetchAuthToken()
+	if err != nil {
+		cancel()
+		err = fmt.Errorf("unable to fetch authentication token: %w", err)
+		log.Infof("unable to establish stream, will possibly retry: %s", err)
+		return nil, err
+	}
+
+	streamCtx, streamCancel := context.WithCancel(
+		metadata.NewOutgoingContext(ctx, metadata.MD{
+			"authorization": []string{fmt.Sprintf("Bearer %s", token)},
+		}),
+	)
+
+	creds := credentials.NewTLS(&tls.Config{
+		InsecureSkipVerify: true,
+	})
+	conn, err := grpc.DialContext(
+		ctx,
+		fmt.Sprintf(":%v", config.Datadog.GetInt("cmd_port")),
+		grpc.WithTransportCredentials(creds),
+	)
+	if err != nil {
+		streamCancel()
+		cancel()
+		return nil, err
+	}
+
+	cas := &coreAgentStream{
+		streamCtx:    streamCtx,
+		streamCancel: streamCancel,
+		cancel:       cancel,
+		agentClient:  pbgo.NewAgentSecureClient(conn),
+		tufClient:    tuf.NewDirectorPartialClient(),
+	}
+
+	cas.connect()
+	return cas, nil
+}
+
+func (cas *coreAgentStream) connect() {
+	cas.mx.Lock()
+	defer cas.mx.Unlock()
+	stream := cas.getStream()
+	cas.stream = stream
+}
```

Comment:
Why do we lock here? 

Should `connect()` and `getStream()` be a single method that always enforces locking? I see we use `getStream()` in other places, is it guaranteed to always return the same stream?

Reply:
It can be called from multiple goroutines. But I refactored this in https://github.com/DataDog/datadog-agent/pull/9939/commits/d624a3911b40ea72b3bb0780f518f3efd67c1a65 anyway, so let me know if it is clearer now.